deepchecks/recommender/ranking.py

```diff
@@ -7,705 +7,428 @@
-    Initialize a weight array consistent of 1s if none is given
-    This is called at the start of each function containing a w param
-    :params:
-    w      : a weight vector, if one was given to the initial function, else None
-             NOTE: w MUST be an array of np.float64. so, even if you want a boolean w,
-             convert it to np.float64 (using w.astype(np.float64)) before passing it to
-             any function
-    n      : the desired length of the vector of 1s (often set to len(u))
-    :returns:
-    w      : an array of 1s with shape (n,) if w is None, else return w un-changed
-    """
-    if w is None:
-        return np.ones(n)
-    else:
-        return w
 
 
-@jit(nopython=True, fastmath=True)
-def cosine(u, v, w=None):
-    """
-    :purpose:
-    Computes the cosine distance between two 1D arrays
-    NOTE: fastdist before v1.1.5 returned cosine similarity (which is 1 - distance)
-    :params:
-    u, v   : input arrays, both of shape (n,)
-    w      : weights at each index of u and v. array of shape (n,)
-             if no w is set, it is initialized as an array of ones
-             such that it will have no impact on the output
-    :returns:
-    cosine  : float, the cosine distance between u and v
-    """
-    n = len(u)
-    w = init_w(w, n)
-    num = 0
-    u_norm, v_norm = 0, 0
-    for i in range(n):
-        num += u[i] * v[i] * w[i]
-        u_norm += abs(u[i]) ** 2 * w[i]
-        v_norm += abs(v[i]) ** 2 * w[i]
-
-    denom = (u_norm * v_norm) ** (1 / 2)
-    return 1 - num / denom
-
-
-@jit(nopython=True, fastmath=True)
-def cosine_vector_to_matrix(u, m):
-    """
-    :purpose:
-    Computes the cosine similarity between a 1D array and rows of a matrix
-    :params:
-    u      : input vector of shape (n,)
-    m      : input matrix of shape (m, n)
-    :returns:
-    cosine vector  : np.array, of shape (m,) vector containing cosine similarity between u
-                     and the rows of m
-    (returns an array of shape (100,))
-    """
-    norm = 0
-    for i in range(len(u)):
-        norm += abs(u[i]) ** 2
-    u_norm = u / norm ** (1 / 2)
-    m_norm = np.zeros(m.shape)
-    for i in range(m.shape[0]):
-        norm = 0
-        for j in range(len(m[i])):
-            norm += abs(m[i][j]) ** 2
-        m_norm[i] = m[i] / norm ** (1 / 2)
-    return np.dot(u_norm, m_norm.T)
-
-
-@jit(nopython=True, fastmath=True)
-def cosine_matrix_to_matrix(a, b):
-    """
-    :purpose:
-    Computes the cosine similarity between the rows of two matrices
-    :params:
-    a, b   : input matrices of shape (m, n) and (k, n)
-             the matrices must share a common dimension at index 1
-    :returns:
-    cosine matrix  : np.array, an (m, k) array of the cosine similarity
-                     between the rows of a and b
-    (returns an array of shape (10, 100))
+def recall_k(relevant_items: Union[List[int], np.array, None],
+             recommendation: Union[List, np.array, None],
```

Comment:
regarding types:
1. Can this be None? In what case?
2. Thought we decided that item could be a int or a str, so should be List[Union[str, int]] no?
3. recommendation is a single item no? Otherwise it should have been recommendations

Same comment for all the metrics here

Reply:
The purpose of the None value is to handle cases where there are no relevant items or no recommendations available for evaluation. In such cases, the function returns a recall score of 0.0.

By allowing None as a valid input, the function provides flexibility to handle different scenarios. For example, if there are no relevant items for a user, it would not be fair to penalize the recommendation system by returning a low recall score. Similarly, if there are no recommendations generated by the system, it would also result in a recall score of 0.0.